File: twitter-backend/src/main/java/com/fei/twitterbackend/model/entitiy/Tweet.java
package com.fei.twitterbackend.model.entitiy;

import jakarta.persistence.*;
import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "tweets")
@Data
public class Tweet {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(length = 280)
    private String content;

    private String mediaUrl;

    // --- REPLY LOGIC (Option A: Flat) ---
    // We only need to know the immediate parent.
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Tweet parentTweet;

    // We don't map the "List<Tweet> replies" here usually,
    // because fetching a tweet shouldn't automatically fetch 10,000 replies.
    // You fetch replies via a separate Repository query.

    // --- RETWEET LOGIC ---
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "retweet_id")
    private Tweet originalTweet;

    // --- COUNTERS ---
    private int replyCount = 0;
    private int retweetCount = 0;
    private int likeCount = 0;

    @CreationTimestamp
    private LocalDateTime createdAt;

    // Helper to check type
    public boolean isRetweet() {
        return originalTweet != null;
    }

    public boolean isReply() {
        return parentTweet != null;
    }
}

File: twitter-frontend/src/types/tweet.ts
import type { UserResponse } from './user';

export interface TweetRequest {
    content?: string;
    parentId?: number | null;
}

export interface TweetResponse {
    id: number;
    content: string | null;
    mediaType: string | null;
    mediaUrl: string | null;
    user: UserResponse;
    replyCount: number;
    likeCount: number;
    retweetCount: number;
    isLiked: boolean;
    isRetweeted: boolean;
    retweetedTweet?: TweetResponse | null;
    replyToTweetId?: number | null;
    replyToUsername?: string | null;
    createdAt: string;
}
